Features/pattern practic

SpectrumDen opened the pull request.

Liqudator committed:
feat: Практическая работа по принципам ООП и паттернам проектирования

Задача: Реализовать приложение для экспорта документа из некоего
хранилища. Документы могут быть как простыми, так и составными
(комплектами). Комплекты – это иерархические структуры, которые могут
включать в себя отдельные документы или другие комплекты.

Так как структура представляет из себя иерархическое дерево, то был
применен паттерн Компоновщик, чтобы работать с деревом как единичный
объект. Узлами которого является Комплект, а конечными элементами -
Документ.

Добавлены сервисы:
1. StorageService - Хранилище, откуда будут доставаться данные.
2. ExportService - Сервис по экспортированию данных.
Для уменьшения связей между ними добавлены контракты сервисов, чтобы
в будущем применить инверсию зависимостей с помощью любого Ioc-
контейнера.

Liqudator committed:
feat: Добавлены смысловые пространства имен

Liqudator committed:
feat: Добавлен сервис по шифрованию

Liqudator committed:
feat: Добавлена сервис по архивации данных

gmike-dev commented on PatternPractic/PatternPractic/Program.cs:
```diff
@@ -0,0 +1,39 @@
+  {
+    public static void Main(string[] args)
+    {
+      var storageService = new StorageService();
```
Попробуй заиспользовать паттерн декоратор. Сейчас ты реализовал экспорт, сжатие и шифрование как отдельные сервисы, и ими управляет экспортер, и экспорт выполняется тремя разными методами. Если экспорт и сжатие реализовать в виде декораторов, можно привести их к одному интерфейсу

coffeek commented on PatternPractic/PatternPractic/Services/EncryptService.cs:
```diff
@@ -0,0 +1,13 @@
+{
+  public class EncryptService : IEncryptService
+  {
+    public void Encrypt(object data)
```
может IComponent?

gmike-dev reviewed (changes requested):
Из всего курса увидел только паттерн компоновщика. Например, для сжатия и шифрования файлов было бы неплохо использовать паттерн декоратор.

Вот это условие из задачи "намекало" на использование ещё одного паттерна:

> Шифрование и экспорт выполняются с использованием сторонних библиотек.

Кстати, в решении реализацию этого условия я не увидел.

Подумай, какие паттерны можно применить в решении и надо доделать

gmike-dev commented on PatternPractic/PatternPractic/Extensions/IntExtensions.cs:
```diff
@@ -0,0 +1,18 @@
+  {
+    public static string GetIndentString(this int indent)
+    {
+      var result = new StringBuilder();
```
`new string(' ', indent)`

gmike-dev commented on PatternPractic/PatternPractic/ArchiveService.cs:
```diff
@@ -0,0 +1,14 @@
+{
+  public class ArchiveService : IArchiveService
+  {
+    public void Archive(object data)
```
Может вместо object-ов везде передавать IComponent? Можно будет подробнее залогировать, какой документ обработан
```suggestion
    public void Archive(IComponent data)

```

gmike-dev commented on PatternPractic/PatternPractic/Domain/Component.cs:
```diff
@@ -0,0 +1,15 @@
+    public int Id { get; set; }
+    public string Name { get; set; }
+
+    public string Description { get; }
```
Из задачи:

>Описание документа - это форматированный в виде дерева текст (с помощью отступов), включающий в себя имя самого документа и документов, входящих в него (на всю глубину иерархии). Имена произвольных документов могут меняться, при этом описание документа должно оставаться в актуальном состоянии (то есть оно не хранится). 

Т.е. для обычного документа Description должен возвращать Name, а для комплекта - к имени достраивать ещё и Description вложенных документов

gmike-dev replied:
Сейчас это просто какое-то строковое свойство, которое нигде не используется

gmike-dev commented on PatternPractic/PatternPractic/Domain/IComponent.cs:
```diff
@@ -0,0 +1,10 @@
+﻿namespace PatternPractic.Domain
+{
+  public interface IComponent
```
Может переименовать в IDocument? Смущает название. Комплекты и документы - это всё документы

gmike-dev commented on PatternPractic/Storage/Document.cs:
```diff
@@ -0,0 +1,19 @@
+﻿using System;
```
Какой-то левый файл?

gmike-dev commented on PatternPractic/Storage/DocumentStorage.cs:
```diff
@@ -0,0 +1,27 @@
+﻿using System;
```
Какой-то левый файл?

gmike-dev reviewed (changes requested):
Поправь ещё по замечаниям выше

Liqudator committed:
Удалил лишние файлы

Liqudator committed:
refactor: Переименовал IComponent на IDocument

Liqudator committed:
Поправил неймспейс сервиса архивации

Liqudator committed:
Добавил editorconfig

Liqudator committed:
feat: Добавил новый проект для экспорта документов

Заюзал паттерны:
- компонощик - для построения зависимости объектов, похожих на древовидные структуры;
- адаптер - для эксопорта и шифрования используются сторонние библиотеки со своими контрактами, чтобы совместить их со своими
использовал адаптер
- декоратор - головной оперцией всего приложения является экспорт, но в приложении она может дополняться как шифрованием, так и архивацией, поэтому был применен паттерн декоратор, который позволяет добавлять новые поведения на лету.

Для удобства развертывания приложения использовал HostedService, а также применил Microsoft Di для разрешения зависимостей.

Классы с суффиксом Adaptee являются чисто имитационными, чтобы моделировать работу сторонних библиотек.
